### PR TITLE
Own hybrid geometry behind a family owner

### PR DESCRIPTION
**Pull request by ricky-chaoju**
Groundwork for the Nemotron-H stack in #644, PR 1 of 4. The hybrid runtime's geometry is currently ten undeclared attributes that lifecycle mutates onto the runner, and `cache_policy` sizes the linear side as `num_layers` minus the SDPA count. That arithmetic breaks as soon as a model has a third kind of layer. On the 52-layer Nemotron checkpoint it reports 46 recurrent layers when only 23 are.

Because of that, this PR moves the geometry behind one owner, a typed layer plan built by `families/gdn.py`, in the same shape as the encoder pooling loaders from #608. Lifecycle shrinks to a four-line route. The loader table comes with the next PR, so lifecycle imports the GDN builder directly for now.

For evidence, the main one is an interleaved A/B on `Qwen/Qwen3.5-0.8B`. Greedy token ids are bit-identical and `per_block_bytes` is identical. `num_blocks` wobbles a few blocks on both arms since overhead is measured at profile time. The full suite gives 1921 passed with 15 new tests, and error messages on malformed configs got stricter, each pinned by a full-string test.

#598 and #607 both touch these hunks, and #607 passes constructor kwargs this PR deletes, so whichever lands second rebases. Happy to be the one who does. The `MambaSpec` shape and the attention-side wiring stay put until the table lands.

### CODE REVIEW

File: vllm_metal/attention/runtime/families/gdn.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    RecurrentStateGeometry,
+    StateFamilySpec,
+)
+
+_INTERVAL_KEY = "full_attention_interval"
+# Scheduler-side mamba caching strategies the GDN state path implements.
+_SUPPORTED_CACHE_MODES = ("none", "align")
+
+
+def build_gdn_hybrid_plan(
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/families/gdn.py:111]: Move the family config and typed model-args parsing above the public builder. `model_args` is the third-party boundary, so validation belongs here, but parse it once into a small `GDNHybridConfig` dataclass instead of repeatedly calling `_require_positive_int()` across the file.

File: vllm_metal/attention/runtime/hybrid_plan.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        cache_idx: int,
+        state_cache: GDNPagedStateCache,
+    ) -> None: ...
+
+    def rebind_state_cache(
+        self, state_cache: GDNPagedStateCache, *, cache_idx: int
+    ) -> None: ...
+
+
+@dataclass(frozen=True, slots=True)
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/hybrid_plan.py:44]: `HybridLayerPlan` should have one source of truth. Today `kinds`, `attention_indices`, and `state_indices` can disagree. Accept layer kinds and derive the index tuples internally, or validate them in `__post_init__`; also use clearer names like `from_layer_kinds()` and `layer_kind()`.

File: vllm_metal/attention/runtime/hybrid.py [hunk 3]
```diff
@@ -79,62 +43,46 @@ class HybridPagedAttentionRuntime(PagedAttentionRuntimeBase):
     def __init__(
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/hybrid.py:43]: Use descriptive ownership names here: `hybrid_plan`, `state_family`, and `state_geometry`. After assigning `self._hybrid_plan`, use that consistently instead of switching between the constructor argument and instance state.

File: vllm_metal/v1/cache_policy.py [hunk 4]
```diff
@@ -67,6 +69,17 @@ def _align_state_pool_count(num_linear_layers: int, num_sdpa_layers: int) -> int
     return num_linear_layers
 
 
+def _require_runner_hybrid_plan(runner: MetalModelRunner) -> HybridRuntimePlan:
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py]: This top-level require helper plus `_require_hybrid_plan()` wrapper is defensive plumbing. Keep the invariant check in one owner, preferably a private `ModelCachePolicy._hybrid_plan()` method, and make the cache path read as: get plan, inspect layer role, emit spec.

File: vllm_metal/v1/cache_policy.py [hunk 5]
```diff
@@ -365,27 +381,33 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
             num_spec_layers, _ = self._runner._yoco_cache_mapping
         specs: dict[str, KVCacheSpec] = {}
         use_deferred_mha_layout = self._uses_deferred_mha_layout()
+        hybrid_plan = self._require_hybrid_plan() if self._runner.is_hybrid else None
         for layer_idx in range(num_spec_layers):
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py]: Compute the layer role once and dispatch clearly. The repeated `kind_of()` calls and impossible fallback make this harder to read. Handle the roles PR 1 supports now; add `stateless` in PR 2 when Nemotron-H has a real consumer.

File: vllm_metal/v1/model_runner.py [hunk 6]
```diff
@@ -459,6 +460,11 @@ def __init__(self, vllm_config: VllmConfig):
         # present: it is consulted on the KV-sizing path, which runs for every
         # model, not only the YOCO ones that install a real mapping.
         self._yoco_cache_mapping: tuple[int, dict[int, int]] | None = None
+        # Hybrid layer split and state family, resolved once by ModelLifecycle
```
Comment (LxYuan0420) [inline vllm_metal/v1/model_runner.py]: This comment is doing too much. The field name should carry most of the meaning; keep one short sentence and avoid adding heavy architecture notes to the runner.

File: tests/attention/test_hybrid_runtime_plan.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        plan=_make_tiny_plan(),
+        max_num_seqs=2,
+        num_kv_heads=1,
+        head_dim=4,
+        block_size=4,
+        dtype=mx.float32,
+    )
+
+
+class TestGdnPlanDecision:
```
Comment (LxYuan0420) [inline tests/attention/test_hybrid_runtime_plan.py:94]: Trim the test scaffolding. Remove `Arrange / Act / Assert` comments and avoid pinning full internal error strings unless the wording is user-facing. These tests should prove topology, spec sizing, and wrapper placement.

**Review by LxYuan0420 — COMMENTED** [review]
This is the right direction, but I would tighten the contract before merging. Moving hybrid geometry into a typed owner is better than lifecycle mutating many runner fields, but the current shape still has inconsistent topology construction, defensive helper plumbing, and naming that makes the flow harder to read.

**Comment by ricky-chaoju** [comment]
Done. I ended up deleting from_kinds rather than renaming it; the constructor takes kinds directly now, so the factory had nothing left to do. Kept the full-string asserts only where a user can actually hit the message; the internal ones match fragments now.

File: vllm_metal/v1/model_runner.py (hunk 6 above)
Comment (ricky-chaoju) [reply vllm_metal/v1/model_runner.py]: ok,will watch it.

File: vllm_metal/attention/runtime/families/gdn.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    def _require_positive_int(model_args: Mapping[str, Any], key: str) -> int:
+        value = model_args.get(key)
+        if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
+            raise ValueError(
+                f"GDN hybrid model args are missing a usable {key!r}: got {value!r}."
+            )
+        return int(value)
+
+
+def build_gdn_hybrid_plan(
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/families/gdn.py:111]: I noticed this still accepts `full_attention_interval=1` and `full_attention_interval > num_layers`, which creates “hybrid” plans with zero state layers or zero SDPA layers instead of failing at the GDN config boundary.

File: vllm_metal/attention/runtime/hybrid_plan.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    def attention_cache_index(self, layer_idx: int) -> int:
+        """Return the compact KV cache ordinal owned by an attention layer."""
+        return self._cache_index(self.attention_indices, layer_idx, ATTENTION_LAYER)
+
+    def state_cache_index(self, layer_idx: int) -> int:
+        """Return the compact state ordinal owned by a state layer."""
+        return self._cache_index(self.state_indices, layer_idx, STATE_LAYER)
+
+    @staticmethod
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/hybrid_plan.py]: `_cache_index()` feels like private helper plumbing inside a small value object. Keep this direct and readable, or expose it as `cache_index()` if it is part of the object contract.

File: vllm_metal/attention/runtime/families/gdn.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    model_args: Mapping[str, Any], num_layers: int
+) -> HybridRuntimePlan:
+    """Resolve GDN layer topology and recurrent geometry from model args."""
+    config = GDNHybridConfig.from_model_args(model_args)
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/families/gdn.py]: Use more descriptive names here: `gdn_config` and `layer_roles` read clearer than `config` and `kinds`. This is the family-owned topology builder, so the variable names should explain the model layout directly.

File: vllm_metal/attention/runtime/families/gdn.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    """GDN dims parsed once from the third-party model args boundary."""
+
+    full_attention_interval: int
+    num_key_heads: int
+    num_value_heads: int
+    key_head_dim: int
+    value_head_dim: int
+    conv_kernel_dim: int
+
+    @classmethod
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/families/gdn.py:39]: This parsing path is still noisier than it needs to be. `model_args` is the boundary, so validation belongs here, but the private static helper makes the dataclass harder to read. Prefer a direct `GDNHybridConfig.from_model_args()` flow, or one small module-level reader if we really need shared parsing.

File: vllm_metal/attention/runtime/hybrid_plan.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+from typing import Any, Literal, Protocol, TypeAlias
+
+import mlx.nn as nn
+import torch
+from vllm.v1.attention.backends.registry import MambaAttentionBackendEnum
+from vllm.v1.kv_cache_interface import MambaSpec
+
+from vllm_metal.attention.caches.gdn_cache import GDNPagedStateCache
+
+LayerKind: TypeAlias = Literal["attention", "state"]
```
Comment (LxYuan0420) [inline vllm_metal/attention/runtime/hybrid_plan.py]: `LayerKind` / `kinds` is still ambiguous. I would call this `LayerRole` with a `layer_roles` field, because consumers are dispatching cache and runtime ownership by layer role.

File: vllm_metal/v1/cache_policy.py [hunk 13]
```diff
@@ -365,25 +376,23 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
             num_spec_layers, _ = self._runner._yoco_cache_mapping
         specs: dict[str, KVCacheSpec] = {}
         use_deferred_mha_layout = self._uses_deferred_mha_layout()
+        hybrid_plan = self._hybrid_plan() if self._runner.is_hybrid else None
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py]: This loop still mixes generic cache policy with hybrid dispatch details. Since `HybridRuntimePlan` owns layer roles, make this read as: get plan, get layer role, emit state spec or attention spec. Avoid repeated `hybrid_plan is not None` checks inside the loop.

File: vllm_metal/v1/cache_policy.py [hunk 14]
```diff
@@ -1292,9 +1292,14 @@ def _hybrid_align_state_bytes_per_block(self) -> int:
             return 0
         if runner.cache_config.mamba_cache_mode != "align":
             return 0
-        num_linear = runner.num_layers - len(runner.sdpa_layer_indices)
-        pools = _align_state_pool_count(num_linear, len(runner.sdpa_layer_indices))
-        return runner.linear_cache_bytes_per_slot() * pools // num_linear
+        # linear_cache_bytes_per_slot routes through ModelCachePolicy, whose
+        # _hybrid_plan() enforces the plan-resolved invariant.
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py]: The comment plus `assert plan is not None` is a smell. If `linear_cache_bytes_per_slot()` already proves the hybrid-plan invariant, keep using the same resolved plan path instead of asserting a second time on `runner.hybrid_runtime_plan`.

File: tests/stub_runner.py [hunk 15]
```diff
@@ -202,3 +212,48 @@ def make_gemma4_mixed_mha_runner(
         head_dim_per_layer=head_dim_per_layer,
         sliding_window_per_layer=sliding_windows,
     )
+
+
+# The production family spec, resolved once through the public builder (the
```
Comment (LxYuan0420) [inline tests/stub_runner.py]: This test helper currently builds a GDN family spec from a degenerate no-state plan. Once the GDN builder rejects invalid hybrid layouts, this should use a valid tiny hybrid config instead of relying on a shape production should not accept.

File: tests/attention/test_hybrid_runtime_plan.py [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+        plan = build_gdn_hybrid_plan(GDN_ARGS, 8)
+
+        assert plan.family.label == "gdn"
+        assert plan.family.wrapper_cls is GDNPagedAttentionWrapper
+        assert plan.family.mamba_type is MambaAttentionBackendEnum.GDN_ATTN
+        assert plan.family.recurrent_dtype is torch.float32
+        assert plan.family.supported_cache_modes == ("none", "align")
+
+
+class TestGdnPlanRejection:
```
Comment (LxYuan0420) [inline tests/attention/test_hybrid_runtime_plan.py:124]: Add focused rejection coverage for `full_attention_interval=1` and `full_attention_interval > num_layers`. Those are the real malformed hybrid cases that still pass today.

File: vllm_metal/v1/cache_policy.py [hunk 17]
```diff
@@ -606,7 +615,7 @@ def _adopt_hybrid_scheduler_group(
             kv_cache_config,
             tuple(
                 f"layers.{layer_idx}.self_attn"
-                for layer_idx in sorted(self._runner.sdpa_layer_indices)
+                for layer_idx in self._hybrid_plan().layers.attention_indices
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py]: `_adopt_hybrid_scheduler_group()` calls `_hybrid_plan()` several times and rebuilds names from the plan each time. Fetch `hybrid_plan` and `layer_plan` once at the top, then use those locals. It keeps the scheduler-layout path easier to audit.

**Review by LxYuan0420 — APPROVED** [review]
Hey @ricky-chaoju, I pushed a few commits to tighten this up before merge. Please review it as well when you have a chance.

**Comment by ricky-chaoju** [comment]
All good. Thanks @LxYuan0420 !
